### PR TITLE
Introducing DEVELOPMENT.md

### LINKED ISSUE #5195: Improving `CONTRIBUTING.md`
Our `CONTRIBUTING.md` is long and extensive, but not very good in my opinion.

My issues with it are:
- It's too long, or at least unfocused. I don't expect that most new contributors read it all.
- It's too technical. It only focuses on code and tools, not on people and processes.

So, what should be in it? Here's a rough outline, heavily inspired by Bevy's contributing guide:
- Introduction.
  - Say thanks for wanting to contribute.
  - Link to discord, good-first-issues and code of conduct
  - The "do not copy GNU code"-warning right at the start
 - Getting oriented (like bevy does it)
 - Design goals (also like bevy)
 - How to help
   - Reporting issues
   - What makes a good actionable issue.
   - What info to include.
   - Writing documentation
   - Writing code
- Rust style guide
  - I love how [nushell does this](https://github.com/nushell/nushell/blob/main/CONTRIBUTING.md#our-rust-style)
  - Should include things like:
    - discourage unsafe and encouraging SAFETY comments
    - discourage panics, unwraps, etc.
    - discourage one-off macros
    - discourage reinventing the wheel
    - encourage a good doc comment style
    - how to write good comments
    - discuss `str` vs `OsStr` vs `Path` etc.
- Git etiquette (like nushell)
- Tools (much like the current section)
- Other implementations (pretty good already)
- License (pretty good already)

Here's some examples of `CONTRIBUTING.md` files I like:
- [rust](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) (short and sweet)
- [nushell](https://github.com/nushell/nushell/blob/main/CONTRIBUTING.md)
- [bevy](https://github.com/bevyengine/bevy/blob/main/CONTRIBUTING.md) (this one in particular is _excellent_)
- [tokio](https://github.com/tokio-rs/tokio/blob/master/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by zhitkoff**
Adding DEVELOPMENT.md with more detailed instruction on setting local development environment

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
I think you might have accidentally included some commits from another branch. Could you remove those?

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:41]: Do we need to mention all of this? If it's a very common pitfall for newcomers I'm all for including it, but it feels maybe a bit too detailed? All bits of information here also need to be kept up to date (i.e. when people should install Visual Studio 2023 instead of 2022). So, just saying "install rust via rustup" should probably be fine. We could also link to more extensive installation guides instead.

This is how simple it could be: https://github.com/nushell/nushell/blob/main/CONTRIBUTING.md#setup (though we could add a bit more info if we really want to).

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:295]: Let's move this to the top.

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:9]: > configured to work with GitHub.

Nit, but the IDE doesn't need GitHub integration.

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:12]:
```suggestion
1. Fork the [coreutils repository](https://github.com/uutils/coreutils) to your GitHub account [ref:](https://docs.github.com/en/get-started/quickstart/fork-a-repo)
```

I'd also like to have some more text as the link than "ref". Maybe `See [this GitHub guide](...) for more information.`

**Review by tertsdiepraam — COMMENTED** [review]
Cool! I think as general feedback I want this to be a bit more concise where possible or at least structure it in a way where people can get started quickly.

Maybe we could also use tricks like the HTML `<details>` element to hide platform-specific instructions?

File: DEVELOPMENT.md
Comment (zhitkoff) [reply DEVELOPMENT.md:41]: True, could be simplified. On the Nu shell example though - they offload all details to linked https://www.nushell.sh/book/installation.html for installing Rust toolchain and all dependencies/prerequisites - that makse their CONTRIBUTING.md clean and simple. Woulkd the equivalent in our case be moving all those details out of CONTRIBUTING.md to DEVELOPMENT.md ?

Comment (tertsdiepraam) [reply DEVELOPMENT.md:41]: Hmm, yeah I guess you're right. We can keep it and just trim it a little bit.

File: DEVELOPMENT.md
Comment (zhitkoff) [reply DEVELOPMENT.md:295]: to the top of the Testing section or somewhere else?

File: DEVELOPMENT.md
Comment (zhitkoff) [reply DEVELOPMENT.md:47]: There are a few things that are needed outside of GNU tests, namely `pre-commin` (even though it is currently broken on MacOS). Also Homebrew seems to be inevitable for development environment on MacOS anyway, and especially so for this project

Comment (tertsdiepraam) [reply DEVELOPMENT.md:48]: But tools like pre-commit also have their own section with installation instructions already. For some others I don't understand what they are used for (like wget & texinfo are those used by GNU?). So maybe we could also add some comments about which packages are used for what.

I do like providing an easy list like this, but maybe that could go at the end and you can link to it here? I think it's quite important that the start of the document is relevant for everyone. What do you think of a structure like this?

```
## Before you start
## Tools
[All tools and cross-platform stuff here]
## Tips for installing on Mac
## Tips for installing on Windows
```

Comment (zhitkoff) [reply DEVELOPMENT.md:48]: makes sense. and yes, most of the Homebrew formulas are needed by GNU build/tests - i just ran it over and over, fixing failures one by one and that is the final list. 
Will update with structure you suggested

**Comment by zhitkoff** [comment]
made changes to DEVELOPMENT.md and simplified CONTRIBUTING.md (offload and link all dev details to DEVELOPMENT.md). There are some TODOs left - looking for feedback

File: DEVELOPMENT.md
Comment (sylvestre) [inline DEVELOPMENT.md:1]:
```suggestion
<!-- spell-checker:ignore reimplementing toybox RUNTEST CARGOFLAGS nextest prereq autopoint gettext texinfo automake findutils shellenv libexec gnubin toolchains -->
```

Comment (zhitkoff) [reply DEVELOPMENT.md]: Thanks! If you do not mind me asking - what are you using for spell checker? The VSCode spellchecker extension did not catch this, even though it is enabled for markdown files by default

**Comment by sylvestre** [comment]
please let me know when it is ready to be reviewed :)

File: CONTRIBUTING.md
Comment (sylvestre) [inline CONTRIBUTING.md:137]: please remove it, we don't do it yet

File: CONTRIBUTING.md
Comment (sylvestre) [inline CONTRIBUTING.md:143]: please fix this one too

File: CONTRIBUTING.md
Comment (sylvestre) [reply CONTRIBUTING.md:137]: ping ?

Comment (zhitkoff) [reply CONTRIBUTING.md:137]: apologies - have been traveling. will do

**Comment by zhitkoff** [comment]
> you know that merging isn't necessary ? ;)
> (esp for doc)

thanks, will keep that in mind in the future.

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:3]:
```suggestion
# Setting up your local development environment
```

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:34]:
```suggestion
***Tip***: You might also need to add 'llvm-tools':
```

It might also be nice to mention what those tools are for. Is code cov?

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:13]: Very nitpicky, but there's a few periods missing at the end of sentences throughout this doc.

File: DEVELOPMENT.md
Comment (tertsdiepraam) [inline DEVELOPMENT.md:93]: I think it'd be great to run markdownlint on this document as well :)

There's a lot of long lines for example, which is fine if you read on GitHub or a MarkDown viewer, but it's nice if it is easy to read in editors too.

**Review by tertsdiepraam — COMMENTED** [review]
This is looking very good! The content is great, I just have few comments on the formatting.

File: DEVELOPMENT.md
Comment (zhitkoff) [reply DEVELOPMENT.md:34]: yep, will update

File: DEVELOPMENT.md
Comment (zhitkoff) [reply DEVELOPMENT.md:93]: I have enabled it to lint automaticaly using VSCode plugin of markdownlint - so far it does not flag anything in the current version of the file. Is there a different markdownlint config that should be used by this project?

File: DEVELOPMENT.md
Comment (sylvestre) [inline DEVELOPMENT.md:282]: it cannot be done on a single line ?

Comment (zhitkoff) [reply DEVELOPMENT.md:282]: it sure can, but how about this version instead?
```
brew install \
  coreutils \
  autoconf \
  gettext \
  wget \
  texinfo \
  xz \
  automake \
  gnu-sed \
  m4 \
  bison \
  pre-commit \
  findutils
```

Comment (sylvestre) [reply DEVELOPMENT.md:282]: ship it :)

**Review by tertsdiepraam — APPROVED** [review]
I'm happy with this! I'll let Sylvestre make the final merge, but it's got my approval. Thanks!

**Pull request merged**
